File: backend/app/core/element_layout.py
"""Versioned, optional layout settings. Legacy profiles need no database migration."""
import math


def bounded(value, low, high, default=0):
    if not isinstance(value, (int, float)) or isinstance(value, bool) or not math.isfinite(value):
        return default
    return round(max(low, min(high, value)))


def normalize_box(element, value):
    return {
        "x": bounded(value.get("x"), -100, 100),
        "y": bounded(value.get("y"), 0, 400),
        "width": bounded(value.get("width"), 260 if element == "frame" else 160, 1040) if value.get("width") else 0,
        "height": bounded(value.get("height"), 0, 1000),
    }


def normalize_layouts(value):
    if not isinstance(value, dict) or value.get("version") not in (1, 2):
        return None
    result = {"version": 2}
    for viewport in ("desktop", "mobile"):
        boxes = value.get(viewport)
        if isinstance(boxes, dict):
            result[viewport] = {}
            for element, box in boxes.items():
                if not isinstance(box, dict) or not (element in ("frame", "discord", "audio") or (element.startswith("widget:") and 7 < len(element) <= 167)):
                    continue
                # Old desktop audio offsets were relative to the entire viewport.
                if value["version"] == 1 and viewport == "desktop" and element == "audio":
                    box = {**box, "x": 0, "y": 0}
                result[viewport][element] = normalize_box(element, box)
    return result


def element_box(settings, element, viewport):
    layouts = normalize_layouts(settings.get("elementLayouts")) or {}
    saved = layouts.get(viewport, {}).get(element)
    if saved is not None:
        return saved
    if element != "frame" or viewport == "mobile":
        return {"x": 0, "y": 0, "width": 0, "height": 0}
    width = settings.get("profileFrameWidth", 430)
    return normalize_box(element, {
        "x": settings.get("profileFrameX", 0) * 2 + {"left": -100, "right": 100}.get(settings.get("cardAlign"), 0),
        "y": max(0, settings.get("profileFrameY", 0) * 4),
        "width": (880 if width == 430 else width) * settings.get("profileFrameScale", 100) / 100,
        "height": settings.get("profileFrameHeight", 0),
    })


def element_style(settings, element):
    values = []
    for viewport in ("desktop", "mobile"):
        box = element_box(settings, element, viewport)
        width = f'{box["width"]}px' if box["width"] else "100%"
        values.extend((f"--{viewport}-width:{width}", f'--{viewport}-height:{box["height"]}px', f'--{viewport}-x:{(box["x"] + 100) / 200}', f'--{viewport}-y:{box["y"]}px'))
    return ";".join(values)

File: frontend/lib/element-layout.ts
import type { CSSProperties } from "react";
import type { ProfileConfig } from "./types";

export type LayoutElement = "frame" | "discord" | "audio" | `widget:${string}`;
export type LayoutViewport = "desktop" | "mobile";
/** Frame x is a percentage of available horizontal travel. Other x values offset
 * within their slot; y, width and height are pixels. Only the frame is resizable. */
export interface ElementBox { x: number; y: number; width: number; height: number }
export interface ElementLayouts {
  version: 1 | 2;
  desktop?: Partial<Record<LayoutElement, ElementBox>>;
  mobile?: Partial<Record<LayoutElement, ElementBox>>;
}
export function constrainBox(id: LayoutElement, box: ElementBox): ElementBox {
  const limit = (n: number, min: number, max: number) => Math.round(Math.min(max, Math.max(min, Number.isFinite(n) ? n : 0)));
  return {
    x: limit(box.x, -100, 100),
    y: limit(box.y, -400, 400),
    width: id === "frame" && box.width ? limit(box.width, 260, 1040) : 0,
    height: id === "frame" ? limit(box.height, 0, 1000) : 0,
  };
}
/** Version 1 audio used viewport coordinates on desktop. Its offsets have no
 * equivalent inside a media slot; retain mobile placement but drop old sizing. */
export function normalizeLayouts(layouts?: ElementLayouts): ElementLayouts {
  const next: ElementLayouts = { version: 2 };
  for (const viewport of ["desktop", "mobile"] as const) {
    const entries = layouts?.[viewport];
    if (!entries) continue;
    next[viewport] = {};
    for (const [key, value] of Object.entries(entries)) {
      if (!value || !(key === "frame" || key === "discord" || key === "audio" || (key.startsWith("widget:") && key.length > 7 && key.length <= 167))) continue;
      const id = key as LayoutElement;
      next[viewport]![id] = constrainBox(id, layouts?.version === 1 && viewport === "desktop" && id === "audio" ? { ...value, x: 0, y: 0 } : value);
    }
  }
  return next;
}
export function updateElementLayout(settings: ProfileConfig["settings"], viewport: LayoutViewport, id: LayoutElement, box: ElementBox): ElementLayouts {
  const layouts = normalizeLayouts(settings.elementLayouts);
  return { ...layouts, [viewport]: { ...layouts[viewport], [id]: constrainBox(id, box) } };
}
export function elementBox(settings: ProfileConfig["settings"], id: LayoutElement, viewport: LayoutViewport): ElementBox {
  const saved = normalizeLayouts(settings.elementLayouts)[viewport]?.[id];
  if (saved) return constrainBox(id, saved);
  if (id !== "frame" || viewport === "mobile") return { x: 0, y: 0, width: 0, height: 0 };
  const width = settings.profileFrameWidth ?? 430;
  return constrainBox(id, {
    x: (settings.profileFrameX ?? 0) * 2 + (settings.cardAlign === "left" ? -100 : settings.cardAlign === "right" ? 100 : 0),
    y: Math.max(0, (settings.profileFrameY ?? 0) * 4),
    width: width === 430 ? 880 : width,
    height: settings.profileFrameHeight ?? 0,
  });
}
export function elementStyle(settings: ProfileConfig["settings"], id: LayoutElement): CSSProperties {
  const style: Record<string, string | number> = {};
  for (const viewport of ["desktop", "mobile"] as const) {
    const box = elementBox(settings, id, viewport);
    style[`--${viewport}-width`] = box.width ? `${box.width}px` : "100%";
    style[`--${viewport}-height`] = `${box.height}px`;
    style[`--${viewport}-x`] = (box.x + 100) / 200;
    style[`--${viewport}-offset-x`] = `${box.x / 2}%`;
    style[`--${viewport}-y`] = `${box.y}px`;
  }
  return style as CSSProperties;
}
